File: services/fae/python/fae/dp.py
"""Differential privacy helpers."""

from __future__ import annotations

import math
import random
from typing import MutableSequence, Sequence


def laplace_noise(scale: float, rng: random.Random) -> float:
    if scale <= 0:
        return 0.0
    u = rng.random() - 0.5
    return -scale * math.copysign(math.log(1 - 2 * abs(u)), u)


def apply_dp_noise(
    values: Sequence[float],
    epsilon: float,
    sensitivity: float = 1.0,
    seed: int = 0,
) -> Sequence[float]:
    """Applies Laplace noise to a sequence deterministically for tests."""

    if epsilon <= 0:
        return list(values)
    scale = sensitivity / epsilon
    rng = random.Random(seed)
    noised: MutableSequence[float] = []  # type: ignore[assignment]
    for value in values:
        noised.append(float(value) + laplace_noise(scale, rng))
    return noised

File: services/fae/go/fae/dp.go
package fae

import (
	"math"
	"math/rand"
)

// ApplyDPLaplace applies Laplace noise with deterministic seed for reproducibility.
func ApplyDPLaplace(values []float64, epsilon, sensitivity float64, seed int64) []float64 {
	if epsilon <= 0 {
		copied := make([]float64, len(values))
		copy(copied, values)
		return copied
	}
	scale := sensitivity / epsilon
	rng := rand.New(rand.NewSource(seed))
	noisy := make([]float64, len(values))
	for i, v := range values {
		u := rng.Float64() - 0.5
		noise := -scale * math.Copysign(math.Log(1-2*math.Abs(u)), u)
		noisy[i] = v + noise
	}
	return noisy
}
